### PR TITLE
Plugin architecture documentation

### PR DESCRIPTION
**Pull request by metacosm**

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,74 @@
+= Halkyon capability plugins architecture
+
+Part of what makes https://github.com/halkyonio/operator[Halkyon] interesting is the capability system. While the capability
```
Comment (cmoulliard) [inline README.adoc:3]: Can we add a link pointing where we cover `capability` within the halkyon operator and explaining what is a capability ?

File: README.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+concept is powerful, it only makes sense if capabilities can be added to Halkyon without requiring to modify its core. The
+goal of this plugin architecture is to make it as easy as possible to extend Halkyon by adding new capabilities as plugins. This
+has also the added advantage of being able to decouple the lifecycles of the operator and that of its plugins, which can evolve
```
Comment (cmoulliard) [inline README.adoc]: Should we use `releases` instead of `lifecyles` ?

File: README.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+separately (as long as API compatibility is maintained, of course).
+
+== Architecture overview
+
+The plugin architecture relies at its core on https://github.com/hashicorp/go-plugin[Hashicorp's go-plugin]. This, in turns,
+means that Halkyon plugins run as separate processes of the operator, relying on RPC communication with the core. A plugin,
+therefore, consists in two parts: a client that runs in the operator process, controlling the lifecycle of and interacting with
```
Comment (cmoulliard) [inline README.adoc]: Can you use a list of bullets to introduce the 2 parts ?

File: README.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+the second part of the plugin, a server running in a separate process, implementing the plugin behavior. However, from a user's
+point of view, much, if not all, of that complexity is hidden. We also made a point to hide that complexity for plugins
+implementors so that it is as easy as possible to create new plugins, without having to worry about the RPC infrastructure.
+While the use of RPC makes it technically possible to write plugins using different programming languages, we focused our
+efforts (and will only document) the use case of a Go-based plugin. Each plugin is compiled into a binary and needs to follow
+some conventions in order to be automatically discoverable and downloadable by the operator.
+
+=== Client
+
+The operator is only marginally aware of plugins: it loads them from a local `plugins` directory which can be (and is, in most
```
Comment (cmoulliard) [inline README.adoc]: Is `marginally` the appropriate adjective to be used ?

File: README.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+instances) seeded by passing a list of comma-separated plugin references, which are downloaded and cached locally. Each binary
+found in the `plugins` directory by the operator is then assumed to be a capability plugin which path is passed to the
+`NewPlugin` function. This function sets the RPC plumbing, in particular, starts the plugin process, opens a client to it and
```
Comment (cmoulliard) [inline README.adoc]: Should be good to provide an example covering what you explain ;-)

File: README.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+registers the metadata of the plugin so that the operator can know which capabilities it provides. All this is executed when the
```
Comment (cmoulliard) [inline README.adoc]: Can you detail what the metadata are please ?

File: README.adoc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+operator starts in its `main` function. From there, the operator is only aware of the plugin when it attempts to create a 
+capability: based on its category and type combination, the operator will look for a plugin providing support for such a pair to
```
Comment (cmoulliard) [inline README.adoc]: Add a bullets list

File: README.adoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+and `ReadyFor` directly:
+
+[source,go]
+----
+type Plugin interface {
+	Name() string
+	GetCategory() halkyon.CapabilityCategory
+	GetTypes() []TypeInfo
+	ReadyFor(owner *halkyon.Capability) []framework.DependentResource
+	Kill()
```
Comment (cmoulliard) [inline README.adoc]: To be describe what is the purpose of each field, with examples ?

File: README.adoc (hunk 1 above)
Comment (metacosm) [reply README.adoc:3]: I'd argue that if you're reading about the capability plugin architecture, you already know what a capability is or you probably don't need to read this :)

**Pull request merged**